File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public  boolean isSame(TreeNode p, TreeNode q){
                if(p==null && q==null)
                    return true;
                if(p==null || q==null)
                    return false;
                return  ( p.val==q.val) && isSame(p.left,q.left)
                        && isSame(p.right,q.right);
                
        }
    public boolean isSubtree(TreeNode s, TreeNode t) {
        //Write code here
            if(t==null) return true;
            if(s==null) return false;

            if(isSame(s,t)) return true;
            return isSubtree(s.left,t) || isSubtree(s.right,t);
    }
}

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False
        return p.val==q.val and self.isSameTree(p.left,q.left) and self.isSameTree(p.right,q.right)
    
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        if not root: return False
        
        if not subRoot or self.isSameTree(root,subRoot): return True
        
        return self.isSubtree(root.left,subRoot) or self.isSubtree(root.right,subRoot)
        